### PR TITLE
options/posix: Implement pwrite

### PR DESCRIPTION
**Pull request by Dennisbonke**
This PR implements `pwrite()` as a sysdep, and adds an implementation of `sys_pwrite()` to the Managarm sysdeps.

Blocked on managarm/managarm#430

### CODE REVIEW

File: options/posix/generic/unistd-stubs.cpp [hunk 1]
```diff
@@ -588,9 +588,19 @@ ssize_t pread(int fd, void *buf, size_t n, off_t off) {
 	return num_read;
 }
 
-ssize_t pwrite(int, const void *, size_t, off_t) {
-	__ensure(!"Not implemented");
-	__builtin_unreachable();
+ssize_t pwrite(int fd, const void *buf, size_t n, off_t off) {
+	ssize_t num_read;
```
Comment (avdgrinten) [inline options/posix/generic/unistd-stubs.cpp:592]: Nitpick: I guess this should be called `num_written`.

File: sysdeps/managarm/generic/file.cpp [hunk 2]
```diff
@@ -3917,6 +3917,73 @@ int sys_pread(int fd, void *buf, size_t n, off_t off, ssize_t *bytes_read) {
...
+	HEL_CHECK(offer->error);
+	HEL_CHECK(send_req->error);
+	HEL_CHECK(imbue_creds->error);
+	HEL_CHECK(recv_resp->error);
+
+	managarm::fs::SvrResponse<MemoryAllocator> resp(getSysdepsAllocator());
+	resp.ParseFromArray(recv_resp->data, recv_resp->length);
+/*	if(resp.error() == managarm::fs::Errors::NO_SUCH_FD) {
+		return EBADF;
+	}else*/
```
Comment (avdgrinten) [inline sysdeps/managarm/generic/file.cpp:3970]: ?

File: options/posix/generic/unistd-stubs.cpp (hunk 1 above)
Comment (Dennisbonke) [reply options/posix/generic/unistd-stubs.cpp:592]: Good catch!

File: sysdeps/managarm/generic/file.cpp [hunk 3]
```diff
@@ -3917,6 +3917,73 @@ int sys_pread(int fd, void *buf, size_t n, off_t off, ssize_t *bytes_read) {
...
+	actions[2].type = kHelActionImbueCredentials;
+	actions[2].flags = kHelItemChain;
+	actions[3].type = kHelActionSendFromBuffer;
+	actions[3].flags = kHelItemChain;
+	actions[3].buffer = const_cast<void *>(buf);
+	actions[3].length = n;
+	actions[4].type = kHelActionRecvInline;
+	actions[4].flags = 0;
+	HEL_CHECK(helSubmitAsync(handle, actions, 5,
+			globalQueue.getQueue(), 0, 0));
```
[the remark this answers is not in the document]
Comment (Dennisbonke) [reply sysdeps/managarm/generic/file.cpp:3952]: Yeah it really should, time to work on my helix_ng skills again.

File: sysdeps/managarm/generic/file.cpp (hunk 2 above)
Comment (Dennisbonke) [reply sysdeps/managarm/generic/file.cpp:3970]: Likely copy paste mistake, I'll investigate and remove it / uncomment it if it is used.

File: sysdeps/managarm/generic/file.cpp [hunk 4]
```diff
@@ -3917,6 +3917,57 @@ int sys_pread(int fd, void *buf, size_t n, off_t off, ssize_t *bytes_read) {
...
+	req.set_size(n);
+	req.set_offset(off);
+
+	frg::string<MemoryAllocator> ser(getSysdepsAllocator());
+	req.SerializeToString(&ser);
+
+	auto [offer, send_req, imbue_creds, send_data, recv_resp] = exchangeMsgsSync(
+		handle,
+		helix_ng::offer(
+			helix_ng::sendBuffer(ser.data(), ser.size()),
```
Comment (Geertiebear) [inline sysdeps/managarm/generic/file.cpp:3939]: Why don't you use ``helix_ng::sendBragiHeadTail`` here?

Comment (Dennisbonke) [reply sysdeps/managarm/generic/file.cpp:3939]: I was under the impression that only works for the new bragi messages? This is still a legacy CntRequest.

Comment (Geertiebear) [reply sysdeps/managarm/generic/file.cpp:3939]: Ah, if it's still a legacy request then this is fine, yes.

Comment (Geertiebear) [reply sysdeps/managarm/generic/file.cpp:3939]: Wait is it? Afaict in ``protocols/fs/`` there is only an ``fs.bragi``.

Comment (Dennisbonke) [reply sysdeps/managarm/generic/file.cpp:3939]: Well it’s part of the big CntRequest thingy. That is legacy no? I assumed the bragi sendHeadTail thing only worked for dedicated request messages?

Comment (Geertiebear) [reply sysdeps/managarm/generic/file.cpp:3939]: No it works fine for this too.

**Review by Geertiebear — APPROVED** [review]
LGTM now.

**Pull request merged**